File: packages/node/src/resources/analytics.ts
import type { MissionInbox } from '../client.js';
import type { ActivityGraph, ActivityGraphParams, AnalyticsOverview } from '../types.js';

/**
 * The `analytics` resource. Access via `mi.analytics`.
 */
export class Analytics {
  /** @internal */
  constructor(private readonly client: MissionInbox) {}

  /** Account-wide overview: this month + last month send counts and domain/mailbox counts. */
  async getOverview(): Promise<AnalyticsOverview> {
    return this.client.request<AnalyticsOverview>({
      method: 'GET',
      path: '/api/analytics/overview',
    });
  }

  /**
   * Time-series data for send/receive activity. Supports hourly/daily/
   * monthly/yearly granularity and per-counter filtering.
   *
   * @example
   * ```ts
   * const graph = await mi.analytics.getActivityGraph({
   *   granularity: 'daily',
   *   startDate: '2026-08-01T00:00:00Z',
   *   endDate: '2026-08-31T23:59:59Z',
   *   counters: ['outgoing', 'bounces'],
   * });
   * ```
   */
  async getActivityGraph(params: ActivityGraphParams): Promise<ActivityGraph> {
    const query: Record<string, unknown> = {
      granularity: params.granularity,
      startDate: params.startDate,
      endDate: params.endDate,
    };
    if (params.counters !== undefined) query.counters = params.counters;
    return this.client.request<ActivityGraph>({
      method: 'GET',
      path: '/api/analytics/activity-graph',
      query,
    });
  }
}

File: packages/java/src/main/java/com/missioninbox/resources/Analytics.java
package com.missioninbox.resources;

import com.missioninbox.http.HttpClient;

import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;

/** The {@code analytics} resource. Access via {@code mi.analytics}. */
public final class Analytics {

    private final HttpClient http;

    public Analytics(HttpClient http) {
        this.http = http;
    }

    @SuppressWarnings("unchecked")
    public Map<String, Object> getOverview() {
        return (Map<String, Object>) http.request("GET", "/api/analytics/overview", null, null);
    }

    /**
     * Time-series data for send/receive activity.
     *
     * @param granularity One of {@code hourly}, {@code daily}, {@code monthly}, {@code yearly}.
     * @param startDate ISO 8601 timestamp.
     * @param endDate ISO 8601 timestamp.
     * @param counters Optional list of {@code incoming}, {@code outgoing}, {@code bounces},
     *                 {@code spam}, {@code held}. Pass {@code null} to include all.
     */
    @SuppressWarnings("unchecked")
    public Map<String, Object> getActivityGraph(
            String granularity, String startDate, String endDate, List<String> counters) {
        Map<String, Object> query = new LinkedHashMap<>();
        query.put("granularity", granularity);
        query.put("startDate", startDate);
        query.put("endDate", endDate);
        if (counters != null && !counters.isEmpty()) query.put("counters", counters);
        return (Map<String, Object>) http.request("GET", "/api/analytics/activity-graph", query, null);
    }

    public Map<String, Object> getActivityGraph(String granularity, String startDate, String endDate) {
        return getActivityGraph(granularity, startDate, endDate, null);
    }
}
